### PR TITLE
BUG: Fix experimental dtype slot numbers

### PR DESCRIPTION
**Pull request by seberg**
Unfortunately, I forgot to move them around when introducing ensure_canonical,
but the code relies on the order for simplicitly currently...

Which doesn't matter, just means that the DType part of the experimental ufunc
API is completely unusable in 1.23 right now.

----

Pretty big ooops by me, since I really thought I could tell peple in my SciPy talk that the super minimal unitdtype example will work with NumPy 1.23... And now I need this (so it will work with `main` only).

### CODE REVIEW

**Comment by seberg** [comment]
Shows that I should move the very minimal prototype into numpy itself soon probably, but I wold prefer not to do it quite now (just due to timing really).

**Comment by mattip** [comment]
Maybe we could add a CI job to run the prototype against HEAD?

**Comment by seberg** [comment]
Yeah, but I need to dumb it down a bit to avoid any use of `unyt`, etc.  (or create a similar prototype example).

**Comment by seberg** [comment]
Ah, sorry would not need it for that.  Also, I just realized we have the `ScaledFloat` thingy.  Probably, I just need to make that example use the experimental API (potentially with some trick?).

That is a bit of annoying, but it failing is pretty meaningless, if it wasn't for the fact that I would like to reference a working version on Friday morning :).

**Comment by mattip** [comment]
3.11-dev is failing on a typing error, @BvB93 any thoughts? That is not the fault of this PR.

**Comment by mattip** [comment]
Maybe it is a regression with the recently released 3.11b4?

**Comment by BvB93** [comment]
> Maybe it is a regression with the recently released 3.11b4?

Seems like https://github.com/python/cpython/pull/93754 is the culprit, I'll create a PR in a bit.

**Comment by BvB93** [comment]
Right, a PR for the 3.11b4 regression has been created: https://github.com/numpy/numpy/pull/21982

**Comment by mattip** [comment]
Thanks @seberg for both the PR and for providing a nice platform to discover beta-4 incompatibilities :)
